Restore RecordNotFound when *_ids= can't find records by ID

domcleal opened the pull request:
### Summary

9c9fb19 changed the behaviour of the _ids= setters for associations to
raise an AssociationTypeMismatch when unknown IDs are given:

```
Class: <ActiveRecord::AssociationTypeMismatch>
Message: <"Developer(#43811860) expected, got NilClass(#16732720)">
```

This restores the original ActiveRecord::RecordNotFound exception with a
much clearer error message:

```
Class: <ActiveRecord::RecordNotFound>
Message: <"Couldn't find all Developers with 'id': (1, -9999) [WHERE \"contracts\".\"company_id\" = ?] (found 1 results, but was looking for 2)">
```

Fixes #25719, restores the 4-2-stable behaviour.
### Other Information

9c9fb19 shipped in Rails 5.0.0 and higher, introduced in #20997 to fix #20995.

domcleal committed:
Restore RecordNotFound when *_ids= can't find records by ID

9c9fb19 changed the behaviour of the _ids= setters for associations to
raise an AssociationTypeMismatch when unknown IDs are given:

    Class: <ActiveRecord::AssociationTypeMismatch>
    Message: <"Developer(#43811860) expected, got NilClass(#16732720)">

This restores the original ActiveRecord::RecordNotFound exception with a
much clearer error message:

    Class: <ActiveRecord::RecordNotFound>
    Message: <"Couldn't find all Developers with 'id': (1, -9999) [WHERE \"contracts\".\"company_id\" = ?] (found 1 results, but was looking for 2)">

Fixes #25719

domcleal commented:
> Pull requests are usually filed against the master branch for this repo, but this one is against 5-0-stable.

I've changed the base branch of this PR to master, but I'll probably submit it for 5-0-stable as well if this is accepted.

matthewd commented:
> `Couldn't find all Developers with 'id'`

This is potentially a lie: the point of 9c9fb19b9eda46a76d7ce4fd6cb9cc94bd965e62 is that we may *not* have been looking by `primary_key`.

domcleal committed:
Add test for collection *_ids= setter when association primary key set

Fixes casting of IDs to the data type of the association primary key,
rather than then the data type of the model's primary key. (Tests use a
string primary key on the association, rather than an int.)

Tests issue #20995

domcleal commented:
> > Couldn't find all Developers with 'id'
> 
> This is potentially a lie: the point of 9c9fb19 is that we may not have been looking by primary_key.

Thanks, very good point. I've updated the PR to pass the key into the `raise..` method to correct the error message when the association primary key's set.

I've also added tests in a second commit to both verify the error message for unknown IDs and test the original change in 9c9fb19. This necessitated another small change to cast IDs to the type of the association primary key, not the model primary key.

domcleal commented:
Rebased the PR to fix merge conflicts.

domcleal commented:
Thanks for reviewing @matthewd.

matthewd commented:
Backported as ba66ed094a5007e0746976781729d5d9e71b8cc9

domcleal commented:
Perfect, cheers.
